schemaregistry/schemaregistry.md

```diff
@@ -0,0 +1,257 @@
+
+## 2.2. Schema and Schema Version
+
+Conceptually, a schema is a description of a data structure. Since data
+structures evolve over time, the schema describing them will also evolve over
+time. Therefore, a schema often has multiple versions.
+
+For simple scenarios, the API allows for version management to be automatic and
+transparent. Whenever a schema is updated, a new version number is assigned and
+prior schema versions are retained. The latest available schema is always the
+default version that is retrieved when the API is given just the schema `id`.
+
+A newer schema version might introduce breaking changes or it might only
+introduce careful changes that preserve compatibility. These strategies are not
+subject of this specification, but the API provides a conflict handling
+mechanism that allows an implementation to reject updates that do not comply
+with a compatibility policy, if one has been implemented.
+
+### 2.2.1. Multi-format schema sets
+
+If the schema format has not been restricted at the schema group level, each
+schema may describe the same data structure for encoding in different formats
+concurrently. That implies that the schema managed by this registry is not
+a single document, but MAY be a set of documents that all describe the exact
+same data structure, each in their own particular syntax and using their own 
+type model.
+
+The schema version therefore relates to this abstract notion of schema and the
+described data structure. All documents coexisting within the same version
+SHOULD describe the exact same data structure.
+
+### 2.2.2. Schema attributes
+
+As per the definition above, the schema object is a collection of versions
+of schema documents. It only has an identifier attribute.
+
+#### id
+
+- Type: `String`
+- Description: Identifies the schema.
+- Constraints:
+  - REQUIRED
+  - MUST be a non-empty string
+  - MUST conform with RFC3986/3.3 `segment-nz-nc` syntax
+  - MUST be unique within the scope of the schema registry
+- Examples:
+  - myschema
+  - my-schema
+
+### 2.2.3 Schema version attributes
+
+A schema version is a document. The "body" of a schema version MAY be a text
+document or binary stream. An implementation SHOULD validate whether a
+schema version is valid according to the rules of its format, for instance
+whether it is a valid Avro schema document when the format is Apache Avro.
+
+Within the scope of the schema set, the version is identified by the combination
+of a version number and an optional format identifier. The schema version MAY
+also have an additional, optional unique identifier within the scope of the
+registry.
+
+#### version
+
+- Type: `Integer`
+- Description: The version of the schema. This is a simple counter and tracks
+  the version in the scope of this schema within the schema group. The schema
+  document MAY indicate a schema that follows a different versioning scheme.
+- Constraints:
+  - REQUIRED
+  - Assigned by server.
```

Comment:
I think I'm on the same track as Ryan, I'm wondering if we need to allow for the author to decide the version string so they can choose a simple int or a semvar pattern.

Reply:
The goal of the server-assigned integer is for versioning not to complicate the API model. With automatic numbers, you can make all updates a plain POST on the schema URI and you can enforce the compatibility rules you want using service-side logic as the update happens.  

Introducing breaking changes should really require the pain of a wholly new schema with its own backcompat versioning sequence.

Semver 1.x, 2.x, 3.x is really better captured by 

/schemagroups/myapp/schemas/foo.1/versions/{n}
/schemagroups/myapp/schemas/foo.2/versions/{n}
/schemagroups/myapp/schemas/foo.3/versions/{n}

than by 

/schemagroups/myapp/schemas/foo/versions/1.1
/schemagroups/myapp/schemas/foo/versions/2.2
/schemagroups/myapp/schemas/foo/versions/3.0

Those things under foo are not the same if they don't describe structurally compatible data.

Reply:
By doing this you're basically asking the impl to be a document management system. Would it be so bad if we left that up to the author of the schema files and they could pretty much pick any URL pattern they wanted (within their permissions/scope)? Meaning, if they PUT over an existing one then it updates it. If they PUT to a new URL them they're creating a new one. Then the impl doesn't need any versioning at all, no saving of history, etc.. that's left up to systems that are built for that kind of thing and the results are pushed here for sharing/viewing.

Reply:
Given the ability and suggestion to semver the `dataschema` or schema `id`, complex versioning schemes within a schema seem less necessary. A server-side check to prevent footguns should be possible even without versions, and this API doesn't actually prevent replacing a specified version with an incompatible document at the same URL (delete the existing version and/or schema, then re-create a schema with the same name and re-insert schemas until the version lines up, but the content doesn't).

Reply:
Just discovered this project. Evaluating whether we could use this instead of writing our own schema registry. We want to be able to use semver to represent versions for a schema, where foo v1.y can be used as-is instead of foo v1.x (for y >= x), but foo v1.x and foo v2 are incompatible representations of the same data (and then maybe there'll be a converter registered in our system, etc.).

It's not clear to me which direction the conversation is leaning, but it'd be really lovely if it was possible to control the version id assigned to a new version of a schema. Otherwise, I mean, we can always add a proxy in front that remaps foo/versions/x.y to foo.x/versions/y, I guess. But it would be simpler if we could just control the version id to be created as we want. This is a very localized configuration point, too.

Reply:
One interesting thing I was thinking about with respect to semver support for versioning, is that compatibility checking could perhaps be customized with a different configuration based on a real understanding of semver.  So if a client tries to update the content of a schema that is currently at version `2.7.3`, uploading content to new version `2.7.4` then that is a patch version release and the compatibility checker can be very strict.  But if the client does the same thing to new version `3.0.0` then the compatibility checker can e.g. allow breaking changes.

Just a thought. :)